cpp/inc/bond/protocol/protobuf_binary_writer.h

```diff
@@ -0,0 +1,579 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#pragma once
+
+#include <bond/core/config.h>
+
+#include "detail/protobuf_utils.h"
+#include "detail/simple_array.h"
+
+#include <bond/stream/output_counter.h>
+
+/*
+    Implements Protocol Buffers binary encoding.
+    See https://developers.google.com/protocol-buffers/docs/encoding for details.
+*/
+
+namespace bond
+{
+    template <typename BufferT>
+    class ProtobufBinaryWriter;
+
+    template <typename BufferT>
+    class ProtobufBinaryReader;
+
+
+    class ProtobufBinaryCounter
+    {
+        template <typename Buffer>
+        friend class ProtobufBinaryWriter;
+
+    private:
+        struct type : OutputCounter // Must be a new type and not an alias.
+        {};
+    };
+
+
+    template <typename BufferT>
+    class ProtobufBinaryWriter
+    {
+        class Pass1
+        {
+        public:
+            Pass1(ProtobufBinaryWriter* writer)
+                : _writer(writer)
+            {}
+
+            ~Pass1()
+            {
+                _writer->_it = nullptr;
+            }
+
+        private:
+            ProtobufBinaryWriter* _writer;
+        };
+
+        using Counter = ProtobufBinaryCounter::type;
+
+    public:
+        using Buffer = BufferT;
+        using Reader = ProtobufBinaryReader<Buffer>;
+        using Pass0 = ProtobufBinaryWriter<Counter>;
+
+        explicit ProtobufBinaryWriter(Buffer& output)
+            : _output{ output },
+              _it{ nullptr }
+        {}
+
+        template <typename T>
+        ProtobufBinaryWriter(Counter& output, const ProtobufBinaryWriter<T>& /*pass1*/)
+            : _output{ output },
+              _it{ nullptr }
+        {}
+
+        bool NeedPass0() const
+        {
+            return _it == nullptr;
+        }
+
+        Pass1 WithPass0(Pass0& pass0)
+        {
+            _it = pass0._lengths.begin();
+            return this;
+        }
+
+        void WriteVersion()
+        {}
+
+        //
+        // Write methods
+        //
+        void WriteStructBegin(const Metadata& /*metadata*/, bool base)
+        {
+            if (base)
+            {
+                detail::proto::NotSupportedException("Inheritance");
+            }
+
+            if (!_fields.empty())
+            {
+                const auto& info = _fields.top(std::nothrow);
+
+                if (info.level > 1)
+                {
+                    detail::proto::NotSupportedException("Container nesting");
```

Comment:
Struct nesting? What's the issue there?

Reply:
Not a struct nesting, but a container (with struct element in this case, e.g. `vector<vector<S>>`).